src/seq_mv/vector.c

```diff
@@ -1137,6 +1270,122 @@ hypre_SeqVectorInnerProd( hypre_Vector *x,
    return result;
 }
 
+/*--------------------------------------------------------------------------
+ * hypre_SeqVectorInnerProdTaggedHost
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+hypre_SeqVectorInnerProdTaggedHost( hypre_Vector  *x,
+                                    hypre_Vector  *y,
+                                    HYPRE_Complex *iprod )
+{
+   HYPRE_Complex *x_data      = hypre_VectorData(x);
+   HYPRE_Complex *y_data      = hypre_VectorData(y);
+   HYPRE_Int     *tags        = hypre_VectorTags(x);
+   HYPRE_Int      num_tags    = hypre_VectorNumTags(x);
+   HYPRE_Int      num_vectors = hypre_VectorNumVectors(x);
+   HYPRE_Int      size        = hypre_VectorSize(x);
+   HYPRE_Int      total_size  = size * num_vectors;
+   HYPRE_Int      i;
+
+#if defined(HYPRE_USING_OPENMP)
+   HYPRE_Int      j, num_threads = hypre_NumThreads();
+   HYPRE_Complex *thread_sums    = hypre_CTAlloc(HYPRE_Complex,
+                                                 num_threads * num_tags,
+                                                 HYPRE_MEMORY_HOST);
+#endif
+
+   /* Initialize result */
+   for (i = 0; i < num_tags; i++)
+   {
+      iprod[i] = 0.0;
+   }
+
+#if defined(HYPRE_USING_OPENMP)
+   #pragma omp parallel private(i, j)
+   {
+      HYPRE_Int      tid = hypre_GetThreadNum();
+      HYPRE_Complex *sum = &thread_sums[tid * num_tags];
+
+      #pragma omp for HYPRE_SMP_SCHEDULE
+      for (i = 0; i < total_size; i++)
+      {
+         sum[tags[i]] += hypre_conj(y_data[i]) * x_data[i];
+      }
+
+      #pragma omp critical
+      {
+         for (j = 0; j < num_tags; j++)
+         {
+            iprod[j] += sum[j];
+         }
+      }
+   }
+#else
+   for (i = 0; i < total_size; i++)
+   {
+      iprod[tags[i]] += hypre_conj(y_data[i]) * x_data[i];
```

Comment:
If none of the tags are zero, then GMRES convergence would be immediate?

Reply:
I don't think so, this is related to the previous question. But just to expand:

https://github.com/hypre-space/hypre/blob/70f3970ba5e57f7167ef4a3c15349b5d19618de5/src/parcsr_mv/par_vector.c#L679-L683

`iprod[0]`, which is used in GMRES, would contain the full inner product

Reply:
Ah, so it works for ParVector.  But SeqVector (what I was looking at) is not consistent.  I think all of the InnerProdTagged() routines should do the same thing, i.e., put the full inner product in position zero.  But maybe I'm still missing something in the code.

Reply:
@rfalgout sorry for the delay. This has been fixed in [80e9eb0](https://github.com/hypre-space/hypre/pull/1254/commits/80e9eb0dc33d0099fd62cc684809590b869154da)